src/sqlfluff/core/parser/segments/keyword.py

```diff
@@ -125,6 +127,7 @@ def match(cls, segments, parse_context):
         we assume that ._template is a r"" string, and is formatted
         for use directly as a regex. This only matches on a single segment.
         """
+        cls_ = parse_context._factory if parse_context._factory else cls
```

Comment:
Why do we need to do this? I'm a little confused.

Reply:
If we do need this little idiom (i.e. `parse_context._factory if parse_context._factory else cls`), then I wonder if we should wrap it as a method because we use it quite a few times.

Reply:
In `master`, we frequently create new classes "on the fly", and we rely on built-in Python functionality to work with them:
* `classmethod`
* `type()` 
* `issubclass()`
* etc.

`classmethod` in particular is problematic now with the introduction of `RawSegmentFactory`. It **acts** like a class, but it's not a class, so it doesn't work with the techniques above. Thus, in `classmethod`s of `RawSegment`, we need a way to pass in the `RawSegmentFactory` as well, so the method can act correctly (because the **type** alone, e.g. `KeywordSegment`, no longer contains all the necessary info).

This bit of code extracts the factory and substitutes it. It may be that in `RawSegment`, we don't need this trickery and should just **always** use `parse_context._factory`.

As I mentioned earlier, we could eventually redesign things more thoroughly so we are not using Python class-related features for parser objects.